### PR TITLE
Support custom types in broker

### PR DESCRIPTION
**Pull request by petuhovskiy**
Old methods are unchanged for backwards compatibility. Added `SafekeeperDiscoveryRequest` and `SafekeeperDiscoveryResponse` types to serve as example, and also as a prerequisite for https://github.com/neondatabase/neon/issues/5471

### CODE REVIEW

File: storage_broker/proto/broker.proto [hunk 1]
```diff
@@ -48,3 +54,61 @@ message TenantTimelineId {
...
+    SAFEKEEPER_DISCOVERY_RESPONSE = 4;
+}
+
+// A message with a type.
+message TypedMessage {
+    MessageType type = 1;
+
+    optional SafekeeperTimelineInfo safekeeper_timeline_info = 2;
+    optional SafekeeperDiscoveryRequest safekeeper_discovery_request = 3;
+    optional SafekeeperDiscoveryResponse safekeeper_discovery_response = 4;
```
Comment (arssher) [inline storage_broker/proto/broker.proto:97]: What's the difference between SafekeeperTimelineInfo and SafekeeperDiscoveryResponse?

Comment (petuhovskiy) [reply storage_broker/proto/broker.proto:97]: SafekeeperDiscoveryResponse has less fields than SafekeeperTimelineInfo.
For example, it doesn't have `peer_horizon_lsn` and other fields, which are not necessary for establishing a connection.

**Comment by petuhovskiy** [comment]
I'm currently refreshing my protobuf knowledge by reading docs https://protobuf.dev/programming-guides/proto3/
I think I'll update this PR a bit today-tomorrow, but general idea will stay the same.

**Comment by petuhovskiy** [comment]
@arssher I've finished patches that I wanted, ready for review now

File: storage_broker/src/metrics.rs [hunk 2]
```diff
@@ -23,3 +23,35 @@ pub static NUM_SUBS_ALL: Lazy<IntGauge> = Lazy::new(|| {
...
+
+pub static PROCESSED_MESSAGES_TOTAL: Lazy<IntCounter> = Lazy::new(|| {
+    register_int_counter!(
+        "storage_broker_processed_messages_total",
+        "Number of messages received by storage broker, before routing and broadcasting"
+    )
+    .expect("Failed to register metric")
+});
+
+pub static BROADCASTED_MESSAGES_TOTAL: Lazy<IntCounter> = Lazy::new(|| {
```
Comment (arssher) [inline storage_broker/src/metrics.rs:35]: s/broadcasted/broadcast

**Review by arssher — APPROVED** [review]
One unobvious gotcha here is that we have single sub channel for ttid for all message types, which means in theory one type can completely suppress others in dense stream. Moreover, I wanted to change DEFAULT_CHAN_SIZE to 1 or just make channel watch instead of broadcast as accumulating old messages pointless and broker eats a lot of memory in busy regions; with that change it would be even more important. Probably ok for now, but consider fixing it.

File: storage_broker/benches/rps.rs [hunk 3]
```diff
@@ -91,15 +94,23 @@ async fn subscribe(client: Option<BrokerClientChannel>, counter: Arc<AtomicU64>,
...
-    });
-    let request = SubscribeSafekeeperInfoRequest {
-        subscription_key: Some(key),
     };
-    let mut stream = client
-        .subscribe_safekeeper_info(request)
+
+    let request = SubscribeByFilterRequest {
+        types: vec![TypeSubscription {
+            r#type: MessageType::SafekeeperTimelineInfo.into(),
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply storage_broker/benches/rps.rs:104]: `type` is a reserved keyword in Rust

File: storage_broker/proto/broker.proto [hunk 4]
```diff
@@ -48,3 +54,55 @@ message TenantTimelineId {
...
+    // at least one type to receive any messages.
+    repeated TypeSubscription types = 1;
+
+    // If set and enabled, subscription will emit messages only for the specified tenant/timeline.
+    optional FilterTenantTimelineId tenant_timeline_id = 2;
+}
+
+enum MessageType {
+    UNKNOWN = 0;
+    SAFEKEEPER_TIMELINE_INFO = 2;
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply storage_broker/proto/broker.proto:80]: Yes, this way it matches the field number from `TypedMessage`

File: storage_broker/src/metrics.rs (hunk 2 above)
Comment (petuhovskiy) [reply storage_broker/src/metrics.rs:35]: Why?

**Comment by petuhovskiy** [comment]
Rebased over main, ready to merge.

File: storage_broker/src/metrics.rs (hunk 2 above)
Comment (arssher) [reply storage_broker/src/metrics.rs:35]: well, 1) another metrics 'BROADCAST_DROPPED_MESSAGES_TOTAL' is called like that 2) broadcast conjugates without 'ed' :)

Comment (petuhovskiy) [reply storage_broker/src/metrics.rs:35]: Oh, irregular verb, I see.

` 1) another metrics ` is like "**broadcast** (process) dropped messages", and this metric is about "the total number of **broadcasted** messages (ever)".

Closing the thread :)

**Pull request merged**